File: ruby/notes/lib/widgets/toolbar.rb
# frozen_string_literal: true

require "plushie"

class Notes
  # Toolbar widget for the top of each view.
  # Pure Ruby composite -- no Rust, no binary rebuild.
  class Toolbar
    include Plushie::Widget
    include Plushie::UI

    widget :toolbar, kind: :widget

    prop :title, :string, default: ""
    prop :show_back, :boolean, default: false
    prop :actions, :any, default: []

    def render(id, props)
      Plushie::UI::Context.push([])
      node = container("#{id}_bar", padding: [8, 16]) do
        row("#{id}_row", spacing: 12, width: "fill") do
          button("back", "<") if props[:show_back]
          text("#{id}_title", props[:title], size: 20)
          space("#{id}_spacer", width: "fill")
          (props[:actions] || []).each do |action_id, label|
            button(action_id, label)
          end
        end
      end
      Plushie::UI::Context.pop
      node
    end
  end
end

File: elixir/notes/lib/notes/widgets/toolbar.ex
defmodule Notes.Widgets.Toolbar do
  @moduledoc """
  Top bar with a title, optional back button, and action buttons.

  Pure Elixir composite widget -- no Rust, no binary rebuild.

  ## Props

  - `title` -- heading text (string)
  - `show_back` -- whether to show the back button (boolean, default false)
  - `actions` -- list of `{id, label}` tuples for action buttons
  """

  use Plushie.Widget

  widget(:toolbar)

  prop(:title, :string, default: "")
  prop(:show_back, :boolean, default: false)
  prop(:actions, :any, default: [])

  def view(id, props) do
    import Plushie.UI

    row padding: {12, 0}, spacing: 12, width: :fill do
      if props.show_back do
        button("back", "\u2190")
      end

      text("#{id}_title", props.title, size: 20)

      space(width: :fill)

      for {action_id, label} <- props.actions do
        button(action_id, label)
      end
    end
  end
end
